validate-data.js

```diff
@@ -0,0 +1,212 @@
+/**
+ * Validates the various dates in data.json
+ *
+ * Assumptions:
+ * - validate-schema.js has has already been run before this script been run
+ * - data for all minors in the current major are included
+ * - data for older minors in previous majors may be truncated
+ * - the final minor with Extended support will always show for previous majors
+ * - only the initial minor for the next unreleased major will be included
+ * - only one unreased major will be included in the data
+ */
+
+const dayjs = require('dayjs');
+const customParseFormat = require('dayjs/plugin/customParseFormat');
+dayjs.extend(customParseFormat);
+const data = require('./data.json');
+
+/**
+ * Add time to an existing dateStr while retaining existing formatting
+ *
+ * @param dataStr - dateStr in either YYYY-MM or YYYY-MM-DD format
+ * @param timeValue - int of the number of timeUnit's to add e.g. 3
+ * @param timeUnit - string of the unit of time e.g. 'day' or 'month'
+ * @param plusOneDay - whether to add +1 day
+ */
+function addTime(dateStr, timeValue, timeUnit, plusOneDay) {
+    const hasDayPart = dateStrHasDayPart(dateStr);
+    const format = hasDayPart ? 'YYYY-MM-DD' : 'YYYY-MM';
+    let dayjsObj = dayjs(dateStr, format);
+    dayjsObj = dayjsObj.add(timeValue, timeUnit);
+    if (plusOneDay) {
+        dayjsObj = dayjsObj.add(1, 'day');
+    }
+    return dayjsObj.format(format);
+}
+
+/**
+ * Whether the dateStr contains a date part or not i.e. if it is in YYYY-MM-DD or YYYY-MM format
+ */
+function dateStrHasDayPart(dateStr) {
+    return dateStr.match(/^\d{4}-\d{2}-\d{2}?$/);
+}
+
+/**
+ * Get the 'major' part of a version
+ */
+function getMajor(obj) {
+    const match = obj.version.match(/^(\d+)\.(\d+)$/);
+    if (!match) {
+        error(`Invalid version: ${obj.version}`);
+    }
+    return match[1] * 1;
+}
+
+/**
+ * Get a specific version obj
+ * Will return null if version does not exist
+ */
+function getVersionObj(versionNumber) {
+    for (const obj of data.data) {
+        if (obj.version === versionNumber) {
+            return obj;
+        }
+    }
+    return null;
+}
+
+/**
+ * Whether obj has been released or not
+ * If releaseDate has a days part i.e. is `YYYY-MM-DD`, then its assumed it has been released
+ */
+function hasBeenReleased(obj) {
+    return dateStrHasDayPart(obj.releaseDate);
+}
+
+/**
+ * Print an error to console and exit with a status code of 1
+ */
+function error(message) {
+    console.error(`${scope} ${version} ${message}`);
+    process.exit(1);
+}
+
+// Globally scoped variables used in error()
+let scope = '';
+let version = '';
```

Comment:
Why globally scoped and not passed into the function?

Reply:
Updated